x/sp/keeper/sp_status.go

```diff
@@ -0,0 +1,125 @@
+package keeper
+
+import (
+	"cosmossdk.io/errors"
+	"github.com/bnb-chain/greenfield/x/sp/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k Keeper) UpdateToInMaintenance(ctx sdk.Context, sp *types.StorageProvider, requestDuration int64) error {
+	params := k.GetParams(ctx)
+	store := ctx.KVStore(k.storeKey)
+	newRecord := &types.MaintenanceRecord{
+		Height:          ctx.BlockHeight(),
+		RequestDuration: requestDuration,
+		RequestAt:       ctx.BlockTime().Unix(),
+	}
+	key := types.GetStorageProviderMaintenanceRecordsKey(sdk.MustAccAddressFromHex(sp.OperatorAddress))
+	bz := store.Get(key)
+	sp.Status = types.STATUS_IN_MAINTENANCE
+	if bz == nil {
+		if requestDuration > params.GetMaintenanceDurationQuota() {
+			return errors.Wrapf(types.ErrStorageProviderStatusUpdateNotAllow, "not enough quota, quota=%d is less than requested=%d", params.GetMaintenanceDurationQuota(), requestDuration)
+		}
+		newStats := types.SpMaintenanceStats{Records: []*types.MaintenanceRecord{newRecord}}
+		store.Set(key, k.cdc.MustMarshal(&newStats))
+		return nil
+	}
+	var stats types.SpMaintenanceStats
+	k.cdc.MustUnmarshal(bz, &stats)
+	size := len(stats.Records)
+	totalUsedTime := int64(0)
+	// should not happen, stats with len 0 records will be deleted
+	//if size == 0 {
+	//	if requestDuration > params.GetMaintenanceDurationQuota() {
+	//		return errors.Wrapf(types.ErrStorageProviderStatusUpdateNotAllow, "not enough quota, quota=%d is less than requested=%d", params.GetMaintenanceDurationQuota(), requestDuration)
+	//	}
+	//	stats.Records = append(stats.Records, newRecord)
+	//	store.Set(key, k.cdc.MustMarshal(&stats))
+	//	return nil
+	//}
+
+	for i := size - 1; i >= 0; i-- {
+		record := stats.Records[i]
+		if record == nil {
+			continue // should not happen
+		}
+		threshHold := record.GetHeight() + params.GetNumOfLockupBlocksForMaintenance()
+		if ctx.BlockHeight() < threshHold {
```

Comment:
I think this only to be checked for the first record, as if the first record pass the check, the following will do as well.